### PR TITLE
Reinforcement Learning DQN

### PR DESCRIPTION
**Pull request by bepo1337**

### CODE REVIEW

File: reinforcement/reinforcement.py [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+    :param next_state: Next state as one hot encoded
+    :param action_index: Index of the action in the output of the Q-Network
+    :param next_state_as_fen: FEN representation of the next state
+    :return:
+    """
+    if reward is not None:
+        current_state_as_csv = ','.join(['%.0f' % num for num in current_state])
+        next_state_as_csv = ','.join(['%.0f' % num for num in next_state])
+
+        concatenated_example = current_state_as_csv + "+" + str(action) + "+" + str(reward) + "+" + next_state_as_csv + "+" + str(action_index) + "+" + next_state_as_fen
```
Comment (Deischox) [inline reinforcement/reinforcement.py:230]: concatenated_example = f"{current_state_as_csv} + {action} + {reward} + {next_state_as_csv} + {action_index} + {ext_state_as_fen}"

File: reinforcement/reinforcement.py [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+        try:
+            with open(csv_file_name, 'a', newline='') as csv_file:
+                if action_index != -1:
+                    csv_file.write(concatenated_example + '\n')
+                    print(f"save state, {action}, {reward}, next state, action_index, next_state_as_fen to {csv_file_name}")
+                else:
+                    "didnt save state because action_index was invalid"
```
Comment (Deischox) [inline reinforcement/reinforcement.py:237]: print missing

File: reinforcement/reinforcement.py [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+            return -5
+        # had a mate against us but did a good move to escape it. Can this ever happen?
+        elif eval_value_before_action < 0:
+            return +1 # was in mate and got better situation
+
+def create_new_example(state, enemy_player, q_net):
+    """
+    Creates a new example from a state, an enemy and the current network
+    :param state: Current state
+    :param enemy_player: Enemey player
```
Comment (Deischox) [inline reinforcement/reinforcement.py:294]: typo

File: reinforcement/reinforcement.py [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+    #save as example
+    save_example(current_state, agent_move, reward, next_state, action_index, state.fen())
+    return state
+
+def create_new_examples(games, turns, q_net):
+    """
+    Creates new examples to serve as training data
+    :param games: Amount of games to be played
+    :param turns: Amount of turns in each game
+    :param q_net: The Network to be used for selectin agent actions
```
Comment (Deischox) [inline reinforcement/reinforcement.py:330]: typo

File: reinforcement/reinforcement.py [hunk 5]
```diff
@@ -0,0 +1,556 @@
...
+        for turn in range(turns):
+            next_state = create_new_example(next_state, enemy_player, q_net)
+            if next_state is None or next_state.is_game_over():
+                print("Game over or no action could be found in reasonable time!")
+                break
+
+def get_number_of_rows_in_training_set():
+    """
+    Determines the number of rows in the training set
+    :return: NUmber of rows in training set
```
Comment (Deischox) [inline reinforcement/reinforcement.py:346]: typo

File: reinforcement/reinforcement.py [hunk 6]
```diff
@@ -0,0 +1,556 @@
...
+    """
+    with open(csv_file_name) as f:
+       return sum(1 for row in f)
+
+def transform_csv_row_into_parts(row):
+    """
+    Splits the row up into: current_state, action, reward, next_state, action_index, next_state_as_fen
+    :param row: Row to  be split
```
Comment (Deischox) [inline reinforcement/reinforcement.py:354]: Leerzeichen zu viel

File: reinforcement/reinforcement.py [hunk 7]
```diff
@@ -0,0 +1,556 @@
...
+        index_of_highest_q_value_in_copy = torch.argmax(copy_of_q_values)
+        #Match element in original q_values and return the index of that element
+        index_in_orginal_q_values = ((q_values == highest_q_value).nonzero(as_tuple=True)[0])
+        #if theres more than 1 element matching, we just take the first one
+        if index_in_orginal_q_values.shape[0] > 1:
+            index_in_orginal_q_values = index_in_orginal_q_values[0]
+        move = map_action_indice_to_move(state, int(index_in_orginal_q_values))
+        move = uci_to_algebraic_notation(move)
+        if move is not None and move in list(state.legal_moves):
+            found_legal_move = True
```
Comment (Deischox) [inline reinforcement/reinforcement.py:453]: delete

File: reinforcement/reinforcement.py [hunk 8]
```diff
@@ -0,0 +1,556 @@
...
+            return move
+        else:
+            copy_of_q_values = torch.cat(
+                [copy_of_q_values[0:index_of_highest_q_value_in_copy], copy_of_q_values[index_of_highest_q_value_in_copy + 1:]])
+
+def select_best_values_for_each_example(predicted_target_values, next_states_as_fen):
+    """
+    For each of the lists of Q-Values (predicted_target_values) it returns the highest Q-Value legal move for each
```
Comment (Deischox) [inline reinforcement/reinforcement.py:461]: Typo
